src/vs/workbench/contrib/positronIPyWidgets/browser/positronIPyWidgetsService.ts

```diff
@@ -61,28 +61,51 @@ export class PositronIPyWidgetsService extends Disposable implements IPositronIP
 		}));
 	}
 
+	override dispose(): void {
+		super.dispose();
+		// Clean up disposables linked to any connected sessions
+		this._sessionToDisposablesMap.forEach(disposables => disposables.dispose());
+	}
+
 	hasInstance(id: string): boolean {
 		return this._notebookInstancesBySessionId.has(id) ||
 			this._consoleInstancesByMessageId.has(id);
 	}
 
+	/**
+	 * Map to disposeable stores for each session. Used to preventing memory leaks caused by
+	 * repeatedly attaching to the same session which can happen in the case of the application
+	 * closing before the session ends
+	 */
+	private _sessionToDisposablesMap = new Map<ILanguageRuntimeSession, DisposableStore>();
```

Comment:
Using a whole `ILanguageRuntimeSession` as a map key might have unintended consequences (these objects refer to sessions but aren't guaranteed unique). How about keying on the session ID?